# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.
from pathlib import Path
import shutil
import pandas as pd


configfile: "config/config.yaml"


SAMPLE_TABLE = pd.read_csv(config["samples"], index_col="sample_name", sep="\t")
SAMPLE_TABLE.index = SAMPLE_TABLE.index.map(str)
SAMPLES = SAMPLE_TABLE.index.values

output_path_dict = {
    "done_files": Path(config["scratch_dir"]) / "done_files",
    "trimmed_reads": Path(config["scratch_dir"]) / "trimmed_reads",
    "concat_gff": {
        "concat_gff_file": Path(config["scratch_dir"])
        / "concat_gff"
        / "concat_gff.gff",
        "concat_gff_mod_file": Path(config["scratch_dir"])
        / "concat_gff"
        / "concat_gff_mod.gff",
    },
    "concat_genome": {
        "concat_genome_file": Path(config["scratch_dir"])
        / "concat_genome"
        / "concat_genome.fna",
        "concat_genome_done": Path(config["scratch_dir"])
        / "concat_genome"
        / "indexed.done",
    },
    "mapped_reads": Path(config["scratch_dir"]) / "mapped_reads",
    "feature_count": Path(config["scratch_dir"]) / "HTseq",
    "genome_index_parent": Path(config["scratch_dir"]) / "genome_index",
    "library_count": Path(config["scratch_dir"])
    / "raw_library_counts"
    / "library_len.tsv",
    "coverage_positions": Path(config["scratch_dir"]) / "coverage_positions",
    "fastqc": Path(config["scratch_dir"]) / "fastqc",
}

results_path_dict = {
    "counts": Path(config["results"]) / "counts.tsv",
    "counts_json": Path(config["results"]) / "counts.json",
    "annotation_json": Path(config["results"]) / "annotations.json",
    "metadata": Path(config["results"]) / "metadata.tsv",
    "config": Path(config["results"]) / "config.yaml",
    "samples": Path(config["results"]) / "samples.tsv",
    "ref_json": Path(config["results"]) / "bio_db_ref_dict.json",
    "json": Path(config["results"]) / f"{config['experiment_name']}.json",
    "comparisons": Path(config["results"]) / "comparisons.json",
    "samples_json": Path(config["results"]) / "samples.json",
    "metadata_json": Path(config["results"]) / "metadata.json",
    "config_json": Path(config["results"]) / "config.json",
}


rule all:
    input:
        list(results_path_dict.values()),


include: "rules/counting_features.smk"
include: "rules/genome_concat.smk"
include: "rules/gff_tools.smk"


if config["aligner"] == "hisat2":

    include: "rules/map_reads_hisat2.smk"


if config["aligner"] == "bwa":

    include: "rules/map_reads_bwa.smk"


if config["aligner"] == "bowtie2":

    include: "rules/map_reads_bowtie2.smk"


include: "rules/post_htseq2_parsing.smk"


include: "rules/run_trim.smk"


include: "rules/samtools.smk"


include: "rules/unzip.smk"


include: "rules/fastqc.smk"
